From 160fd2dab3977fa9fdf727cb74e6a23b8fc9d20b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor@gmail.com>
Date: Thu, 16 Apr 2015 10:34:57 +0200
Subject: Give priority to rar and unrar that supports RAR files better

rar and unrar supports the RAR5 archive format, while libarchive and p7zip not.

diff --git a/src/fr-archive-libarchive.c b/src/fr-archive-libarchive.c
index 9674098..2b70528 100644
--- a/src/fr-archive-libarchive.c
+++ b/src/fr-archive-libarchive.c
@@ -137,6 +137,16 @@ fr_archive_libarchive_get_capabilities (FrArchive  *archive,
 		return capabilities;
 	}
 
+	/* give priority to rar and unrar that supports RAR files better. */
+	if ((strcmp (mime_type, "application/x-rar") == 0)
+	    || (strcmp (mime_type, "application/x-cbr") == 0))
+	{
+		if (_g_program_is_available ("rar", check_command)
+		    || _g_program_is_available ("unrar", check_command)) {
+			return capabilities;
+		}
+	}
+
 	/* tar.lrz format requires external lrzip */
 	if (strcmp (mime_type, "application/x-lrzip-compressed-tar") == 0) {
 		if (!_g_program_is_available ("lrzip", check_command))
diff --git a/src/fr-command-7z.c b/src/fr-command-7z.c
index e6e8258..62a8fbe 100644
--- a/src/fr-command-7z.c
+++ b/src/fr-command-7z.c
@@ -616,7 +616,10 @@ fr_command_7z_get_capabilities (FrArchive  *archive,
 		if (_g_mime_type_matches (mime_type, "application/x-rar")
 		    || _g_mime_type_matches (mime_type, "application/x-cbr"))
 		{
-			if (! check_command || g_file_test ("/usr/lib/p7zip/Codecs/Rar29.so", G_FILE_TEST_EXISTS))
+			/* give priority to rar and unrar that supports RAR files better. */
+			if (!_g_program_is_available ("rar", check_command)
+			    && !_g_program_is_available ("unrar", check_command)
+			    && g_file_test ("/usr/lib/p7zip/Codecs/Rar29.so", G_FILE_TEST_EXISTS))
 				capabilities |= FR_ARCHIVE_CAN_READ;
 		}
 		else
-- 
cgit v0.10.2

